torch/csrc/api/include/torch/detail/any.h

```diff
@@ -0,0 +1,138 @@
+  Any(const Any& other) = delete;
+  Any& operator=(const Any& other) = delete;
+
+  /// Returns the value contained in the `Any` if the type passed as template
+  /// parameter matches the type of the object stored, and returns a null
+  /// pointer otherwise.
+  template <typename T>
+  T* try_get() {
+    static_assert(
+        !std::is_reference<T>::value,
+        "Any stores decayed types, you cannot cast it to a reference type");
+    static_assert(
+        !std::is_array<T>::value,
+        "Any stores decayed types, you must cast it to T* instead of T[]");
+    if (typeid(T).hash_code() == type_info().hash_code()) {
+      return &static_cast<Holder<T>&>(*content_).value;
+    }
+    return nullptr;
+  }
+
+  template <typename T>
+  T get() {
+    if (auto* value = try_get<T>()) {
+      return *value;
+    }
+    AT_ERROR(
+        "Attempted to cast Any to",
+        at::demangle(typeid(T).name()),
+        ", but its contained type is ",
+        at::demangle(type_info().name()));
+  }
+
+  /// Returns the `type_info` object of the contained value.
+  const std::type_info& type_info() const noexcept {
+    return content_->type_info;
+  }
+
+ private:
+  /// Constructs the `Any` from any type.
+  template <typename T>
+  explicit Any(T&& value)
+      : content_(
+            torch::make_unique<Holder<decay_t<T>>>(std::forward<T>(value))) {}
+
+  /// `Any` is a public type, but its construction is only allowed via this
+  /// "private" function in the `detail` namespace.
+  template <typename T>
+  friend Any detail::make_any(T&& value);
+
+  /// The static type of the object we store in the `Any`, which erases the
+  /// actual object's type, allowing us only to check the `type_info` of the
+  /// type stored in the dynamic type.
+  struct Placeholder {
+    explicit Placeholder(const std::type_info& type_info_) noexcept
+        : type_info(type_info_) {}
+    virtual ~Placeholder() = default;
+    const std::type_info& type_info;
+  };
+
+  /// The dynamic type of the object we store in the `Any`, which hides the
+  /// actual object we have erased in this `Any`.
+  template <typename T>
+  struct Holder : public Placeholder {
+    /// A template because T&& would not be universal reference here.
+    template <typename U>
+    explicit Holder(U&& value_) noexcept
+        : Placeholder(typeid(T)), value(std::forward<U>(value_)) {}
+
+    T value;
+  };
+
+  /// The type erased object.
+  std::unique_ptr<Placeholder> content_;
+};
+
+namespace detail {
+/// Constructs a new `Any` object from any value.
+template <typename T>
+Any make_any(T&& value) {
+  return Any(std::forward<T>(value));
+}
+
+/// A group of methods to construct an `std::vector<Any>` from a variadic list
+/// of arguments.
+inline void push_any(std::vector<Any>& vector) {}
+
+template <typename Head, typename... Tail>
+void push_any(std::vector<Any>& vector, Head&& head, Tail&&... tail) {
+  vector.push_back(make_any(std::forward<Head>(head)));
```

Comment:
This needs to move the any. An implementation with `emplace_back` might be able to avoid that.

Reply:
Actually the `Any` constructor is private